packages/hdwallet-phantom/src/bitcoin.ts

```diff
@@ -0,0 +1,180 @@
+    nonWitnessUtxo?: Buffer;
+    witnessUtxo?: { script: Buffer; value: number };
+  } = {
+    hash: Buffer.from(input.txid, "hex"),
+    index: input.vout,
+  };
+
+  if (input.sequence !== undefined) {
+    inputData.sequence = input.sequence;
+  }
+
+  if (input.scriptType) {
+    switch (input.scriptType) {
+      case "p2pkh":
+        inputData.nonWitnessUtxo = Buffer.from(input.hex, "hex");
+        break;
+      case "p2sh-p2wpkh":
+      case "p2wpkh": {
+        const inputAddress = await wallet.btcGetAddress({ addressNList: input.addressNList, coin, showDisplay: false });
+
+        if (!inputAddress) throw new Error("Could not get address from wallet");
+
+        inputData.witnessUtxo = {
+          script: bitcoin.address.toOutputScript(inputAddress, network),
+          value: parseInt(input.amount),
+        };
+        break;
+      }
+      default:
+        throw new Error(`Unsupported script type: ${input.scriptType}`);
+    }
+  }
+
+  psbt.addInput(inputData);
+}
+
+async function addOutput(
+  wallet: core.BTCWallet,
+  psbt: bitcoin.Psbt,
+  output: core.BTCSignTxOutput,
+  coin: string
+): Promise<void> {
+  if ("address" in output && output.address) {
+    psbt.addOutput({
+      address: output.address,
+      value: parseInt(output.amount),
+    });
+  } else if ("addressNList" in output && output.addressNList) {
+    const outputAddress = await wallet.btcGetAddress({ addressNList: output.addressNList, coin, showDisplay: false });
+
+    if (!outputAddress) throw new Error("Could not get address from wallet");
+
+    psbt.addOutput({
+      address: outputAddress,
+      value: parseInt(output.amount),
+    });
+  } else if ("opReturnData" in output && output.opReturnData) {
+    const data = Buffer.from(output.opReturnData.toString(), "hex");
+    const embed = bitcoin.payments.embed({ data: [data] });
+    psbt.addOutput({
+      script: embed.output!,
+      value: 0,
+    });
+  }
+}
+
+export async function bitcoinSignTx(
+  wallet: core.BTCWallet,
+  msg: core.BTCSignTx,
+  provider: any
+): Promise<core.BTCSignedTx | null> {
+  if (!msg.inputs.length || !msg.outputs.length) {
+    throw new Error("Invalid input: Empty inputs or outputs");
+  }
+
+  const network = getNetwork(msg.coin);
+
+  const psbt = new bitcoin.Psbt({ network });
+
+  psbt.setVersion(msg.version ?? 2);
+  if (msg.locktime) {
+    psbt.setLocktime(msg.locktime);
+  }
+
+  for (const input of msg.inputs) {
+    await addInput(wallet, psbt, input, msg.coin, network);
+  }
+
+  for (const output of msg.outputs) {
+    await addOutput(wallet, psbt, output, msg.coin);
+  }
+
+  const inputsToSign = await Promise.all(
+    msg.inputs.map(async (input, index) => {
+      const address = await wallet.btcGetAddress({
+        addressNList: input.addressNList,
+        coin: msg.coin,
+        showDisplay: false,
+      });
+
+      return {
+        address,
+        signingIndexes: [index],
+        sigHash: bitcoin.Transaction.SIGHASH_ALL,
```

Comment:
I'm not 100% sure about this part tbh, I don't have all the ramifications needed to understand it but what I understand so far is that we want to sign all outputs and input

Reply:
@kaladinlight ^^ confirmed inputs/outputs are happy with latest diff so think we're good here